### PR TITLE
Live Preview: Introduce the upgrade modal in Premium and WooCommerce themes

### PR DESCRIPTION
**Pull request by okmttdhr**
Related to https://github.com/Automattic/wp-calypso/issues/79223, https://github.com/Automattic/wp-calypso/pull/83630, https://github.com/Automattic/wp-calypso/pull/84676
https://github.com/Automattic/wp-calypso/pull/84483

## Proposed Changes

This PR introduces the upgrade modal that activates when a user, while live-previewing a theme, clicks on the `Upgrade` button. 

### UI

the modal (for Premium themes)
Screenshot 2023-12-14 at 15 28 03

the modal (for WooCommerce themes)
Screenshot 2023-12-14 at 15 27 23

'edit' mode
Screenshot 2023-12-11 at 16 01 41 copy

'view' mode
Screenshot 2023-12-11 at 15 14 05

### Flow

https://github.com/Automattic/wp-calypso/assets/5287479/bbe73c28-8435-4ddd-9968-e1ac33bebe2c

## Testing Instructions

- Run `install-plugin.sh wpcom-block-editor add/live-preview-upgrade-modal` on your sandbox.
- Sandbox widgets.wp.com and your site.
- Follow the testing instruction in https://github.com/Automattic/wp-calypso/issues/79223

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [x] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [x] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [x] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/components/theme-upgrade-modal/index.tsx
Comment (okmttdhr) [inline client/components/theme-upgrade-modal/index.tsx:67]: I've adjusted the `ThemeUpgradeModal` for its integration in the Site Editor. I'm open to suggestions if there are more optimal methods to use it outside of Calypso.

File: apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx
Comment (okmttdhr) [inline apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx]: I followed the existing code for other apps and could not delve deeper to find a better way. 🤔 

- https://github.com/Automattic/wp-calypso/blob/373449edcf24f6488de2e084b3595b58a8aebfd4/apps/blaze-dashboard/README.md?plain=1#L30-L41
- https://github.com/Automattic/jetpack/blob/5e1e8c83d06ac24ccc88afdcc386c54d61242709/projects/packages/blaze/src/class-dashboard.php#L83-L89
- https://github.com/Automattic/wp-calypso/blob/373449edcf24f6488de2e084b3595b58a8aebfd4/apps/blaze-dashboard/webpack.config.js#L187-L190

**Comment by autumnfjeld** [comment]
The opacity or the color in the background here looks odd, like it is too white.    Can a different base color be used? Or something changed so the modal pops more against the background, particularly when the theme beneath is white.  

image


Compare to the background here, which appears to me darker more grey.
image

**Comment by okmttdhr** [comment]
> The opacity or the color in the background here looks odd, like it is too white. Can a different base color be used? Or something changed so the modal pops more against the background, particularly when the theme beneath is white.

I fixed it to use the same background color 👍 
Screenshot 2023-12-12 at 15 38 31
Screenshot 2023-12-12 at 15 39 20

**Comment by arthur791004** [comment]
hmmm... I can not see the SVG icons. Am I missing something? But the current way to load SVG icons seems to be fragile...

image

How about replacing icons in the modal with the WordPress icon if possible?

File: apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx
Comment (fushar) [reply apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx]: @okmttdhr We can just inline the SVG code. See: https://github.com/Automattic/wp-calypso/pull/82787

cc: @arthur791004 

Sorry for the late review!

File: client/my-sites/theme/hooks/use-bundle-settings.tsx
Comment (renatho) [inline client/my-sites/theme/hooks/use-bundle-settings.tsx]: Hi @okmttdhr! Thank you for adding me as a reviewer here.
We did a similar thing here in parallel. https://github.com/Automattic/wp-calypso/pull/84886/files#r1422629619

Basically, in my case, I used the `useBundleSettings` getting by the `themeSoftware`, and I created another one with the name `useBundleSettingsByTheme` to get by theme.

I also changed the default export to a named export because I noticed later that it's the way it's used in the majority of the hooks in Calypso.

WDYT? I think it would be good if we apply the same code in both PRs to avoid issues.

File: apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx
Comment (okmttdhr) [reply apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx]: Yes, but I was concerned that we need to discontinue using Gridicons in `client/components/theme-upgrade-modal` (maybe potentially other components in `client/components`). I wanted to keep the component as is as possible. 
But I'm starting to feel it should be acceptable in this specific case.

File: client/my-sites/theme/hooks/use-bundle-settings.tsx
Comment (okmttdhr) [reply client/my-sites/theme/hooks/use-bundle-settings.tsx]: Thank you for letting me know, @renatho! 😄
I will modify my approach to align with your https://github.com/Automattic/wp-calypso/pull/84886, as I prefer `useBundleSettings` if we are retrieving `bundleSettings` from `themeSoftware` on a one-to-one basis.

**Comment by taipeicoder** [comment]
Minor detail, but is it an easy change to update the tooltip text as well?

Screenshot 2023-12-13 at 9 24 42 AM
Screenshot 2023-12-13 at 9 24 25 AM

**Comment by taipeicoder** [comment]
@okmttdhr would it be possible to override the keyboard shortcut for saving? <kbd>⌘</kbd><kbd>S</kbd>
Otherwise, it's possible to work around the Upgrade button:

https://github.com/Automattic/wp-calypso/assets/797888/5520c226-9a4a-4104-bb27-1cfdcff9cc13

File: apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx
Comment (arthur791004) [reply apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx]: >  We can just inline the SVG code

If so, we have to modify the `ThemeUpgradeModal` to use the inline SVG but it seems weird as the component is located under the `client/components` folder and it should use `GridIcon` directly 😅

**Comment by fushar** [comment]
Wait a minute... So if I made some changes, then click `Upgrade now`, I will be redirected to the checkout, then got redirected back to the Site Editor, but my changes are lost... is this the expected flow? 🥲 

I think so because we currently don't have the mechanism to save without activating (CMIIW).

But this seems weird...

**Comment by okmttdhr** [comment]
> Wait a minute... So if I made some changes, then click `Upgrade now`, I will be redirected to the checkout, then got redirected back to the Site Editor, but my changes are lost... is this the expected flow? 🥲
> 
> I think so because we currently don't have the mechanism to save without activating (CMIIW).
> 
> But this seems weird...

Yes, that is expected, and we have no way to save before activation (yet) ... （´・ω・｀）.;:…（´・ω...:.;::..（´・;::: .:.;:

**Comment by fushar** [comment]
I saw https://github.com/Automattic/wp-calypso/pull/85013/commits/8f097602348e4ec7eaa55a1aaa244db4864de24f but I'm still able to "save" the changes using Cmd-S (it's not being overridden), are you sure it works on your side?

**Comment by taipeicoder** [comment]
> > Wait a minute... So if I made some changes, then click `Upgrade now`, I will be redirected to the checkout, then got redirected back to the Site Editor, but my changes are lost... is this the expected flow? 🥲
> > I think so because we currently don't have the mechanism to save without activating (CMIIW).
> > But this seems weird...
> 
> Yes, that is expected, and we have no way to save before activation (yet) ... （´・ω・｀）.;:…（´・ω...:.;::..（´・;::: .:.;:

This seems to be a pretty significant caveat. cc: @autumnfjeld 
Ideally the changes are saved before leaving the Site Editor to the checkout, can we create a spike for this?

File: apps/wpcom-block-editor/src/wpcom/features/live-preview/hooks/use-previewing-theme.ts
Comment (fushar) [inline apps/wpcom-block-editor/src/wpcom/features/live-preview/hooks/use-previewing-theme.ts]: [nit] I believe `previewingThemeSlug` needs not be a dependency here.

**Comment by fushar** [comment]
> hmmm... I can not see the SVG icons. Am I missing something?

Same, I can't see the SVG icons 🥲

File: apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx
Comment (fushar) [inline apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx]: This loads 404 for me (sandboxed):

image

File: apps/wpcom-block-editor/src/wpcom/features/live-preview/hooks/use-override-save-button.ts
Comment (fushar) [inline apps/wpcom-block-editor/src/wpcom/features/live-preview/hooks/use-override-save-button.ts]: Can this be something like

```
tooltip.textContent = __( 'Upgrade now', 'wpcom-live-preview' ) + '⌘S';
```

So that it will just reuse the same translation for the Save button?

Comment (taipeicoder) [reply apps/wpcom-block-editor/src/wpcom/features/live-preview/hooks/use-override-save-button.ts]: I just checked, and LTR locales will have the command on the left side:
Screenshot 2023-12-13 at 4 18 48 PM

Comment (fushar) [reply apps/wpcom-block-editor/src/wpcom/features/live-preview/hooks/use-override-save-button.ts]: (╯°□°)╯︵ ┻━┻

File: client/components/theme-upgrade-modal/index.tsx
Comment (fushar) [reply client/components/theme-upgrade-modal/index.tsx:67]: Should we add a comment saying that we're not supposed to use Calypso state in this component due to this reason?

File: apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx
Comment (okmttdhr) [reply apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx]: I had a try with replacing icons on build time. WDYT? https://github.com/Automattic/wp-calypso/pull/85013/commits/92534977356e9db89ca9129b61818141424903e8

Comment (fushar) [reply apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx]: Sorry maybe I'm missing something but why not just replace gridicons with WP icons in that `ThemeUpgradeModal` directly?

Comment (okmttdhr) [reply apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx]: To be sure, did you see the comment above? 👀 
https://github.com/Automattic/wp-calypso/pull/85013#discussion_r1424813237

Comment (fushar) [reply apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx]: Yes I did. I mean, I don't see why we can't use `@wordpress/icons` in the `ThemeUpgradeModal` if it renders essentially the same thing (a checkmark).

Comment (arthur791004) [reply apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx]: > I don't see why we can't use @wordpress/icons in the ThemeUpgradeModal if it renders essentially the same thing (a checkmark).

+1. I remember I left the same comment somewhere but I cannot see it 😂

Comment (arthur791004) [reply apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx]: > How about replacing icons in the modal with the WordPress icon if possible?

Ooops this

Comment (okmttdhr) [reply apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx]: > replace gridicons with WP icons in that ThemeUpgradeModal directly

Thanks, I’ll address it in that way! 👍 

While I agree that it's acceptable in this specific case, I just want to note that my concern was primarily about replacing the original icons (Gridicons), not specifically about whether to implement them inline. https://github.com/Automattic/wp-calypso/pull/85013#discussion_r1424667153

I've now realized that in Calypso, several different icon sets are being used (`@wordpress/icons`, `Gridicons`, `MaterialIcon`). This is not ideal, badge. Icons are not something interchangeably replaced with different sets even if they appear "similar", and are supposed to be integral to our brand identity.
After things have settled on my end, I’ll create an issue to make it visible.

**Comment by okmttdhr** [comment]
> I saw [8f09760](https://github.com/Automattic/wp-calypso/commit/8f097602348e4ec7eaa55a1aaa244db4864de24f) but I'm still able to "save" the changes using Cmd-S (it's not being overridden), are you sure it works on your side?

There seems to be regression. I'll fix it, and thanks for being patient 😣

File: apps/wpcom-block-editor/src/wpcom/features/live-preview/hooks/use-override-save-button.ts
Comment (okmttdhr) [reply apps/wpcom-block-editor/src/wpcom/features/live-preview/hooks/use-override-save-button.ts]: @taipeicoder @fushar 
I've noticed the command remains on the right side (despite the weird `margin-right`) for the words that have not yet been translated into RTL languages even if I set my language to "العربية". 
Screenshot 2023-12-14 at 12 25 46
Screenshot 2023-12-14 at 12 28 29

But the DOM structure seems to be the same.
Screenshot 2023-12-14 at 12 30 02
Screenshot 2023-12-14 at 12 30 22

Can you help me find out how it's implemented (possibly) on the Gutenberg side? 
We may be able to implement it like `textContent = isRtl ? command + text : text + command` as we override the copy anyway, but I want to understand the current implementation to know if we can just use the current DOM structure without doing `textContent = isRtl ? command + text : text + command`. 

Or would it be okay if we take care of RTL in later iterations? What do you think?

Comment (fushar) [reply apps/wpcom-block-editor/src/wpcom/features/live-preview/hooks/use-override-save-button.ts]: I found that the shortcut text and its placement is computed dynamically. See:

https://github.com/WordPress/gutenberg/blob/2fb3c57bcadc675f9c6745a99d957b3924e39b71/packages/edit-site/src/components/save-button/index.js#L96

My suggestion for now is to NOT include / leave out the shortcut from the tooltip, due to the complexity. What do you think?

Comment (okmttdhr) [reply apps/wpcom-block-editor/src/wpcom/features/live-preview/hooks/use-override-save-button.ts]: > leave out the shortcut from the tooltip

@fushar In that case, the tooltip displays the same text as the button label, making it redundant imo. 😃 
What do you think about hiding the tooltip entirely, then?

Comment (taipeicoder) [reply apps/wpcom-block-editor/src/wpcom/features/live-preview/hooks/use-override-save-button.ts]: I'm in favor of leaving out the shortcut, but keeping the tooltip for accessibility purposes. Even though it might be redundant 🙂

Comment (okmttdhr) [reply apps/wpcom-block-editor/src/wpcom/features/live-preview/hooks/use-override-save-button.ts]: I addressed this in https://github.com/Automattic/wp-calypso/pull/85013/commits/3af814b75852537d84da08d3f53853af25b4c759 🙂

**Comment by okmttdhr** [comment]
I have addressed all the critical feedback. I'd appreciate it if you take a look at it again 🙏 🙏 @Automattic/lego 

- Changed to use WP icons to show them outside of Calypso. https://github.com/Automattic/wp-calypso/pull/85013/commits/be7ac439d3e44c1198f56cdd3af1620a9167f118
- Overrode the default behavior of the keyboard shortcut and the tooltip. https://github.com/Automattic/wp-calypso/pull/85013/commits/b169c3c094f1693522f202f89b27592ed659a067 https://github.com/Automattic/wp-calypso/pull/85013/commits/5ea0caf651601be9562aa5e80c339a190b6d8df1

File: client/my-sites/theme/hooks/use-bundle-settings.tsx
Comment (okmttdhr) [reply client/my-sites/theme/hooks/use-bundle-settings.tsx]: 📝 I tried to be minimalist in applying #84886: https://github.com/Automattic/wp-calypso/pull/85013/commits/d87ac5485d810bad40c7602a25560ad6a9bdfb07

File: apps/wpcom-block-editor/webpack.config.js
Comment (taipeicoder) [inline apps/wpcom-block-editor/webpack.config.js:66]: @okmttdhr could you explain why do we need this change? Thanks!

File: apps/wpcom-block-editor/src/wpcom/features/live-preview/hooks/use-override-save-button.ts
Comment (taipeicoder) [inline apps/wpcom-block-editor/src/wpcom/features/live-preview/hooks/use-override-save-button.ts]: Should this be just:
```suggestion
		document.addEventListener( 'keydown',  overrideSaveButtonKeyboardShortcut );
```

File: client/components/theme-upgrade-modal/index.tsx
Comment (taipeicoder) [inline client/components/theme-upgrade-modal/index.tsx:92]: Should this be `useBundleSettingsByTheme`?

**Review by taipeicoder — APPROVED** [review]
Tested and it works as intended 👍 
Left some minor questions.

File: apps/wpcom-block-editor/src/wpcom/features/live-preview/hooks/use-override-save-button.ts
Comment (okmttdhr) [reply apps/wpcom-block-editor/src/wpcom/features/live-preview/hooks/use-override-save-button.ts]: 🤦

File: client/components/theme-upgrade-modal/index.tsx
Comment (okmttdhr) [reply client/components/theme-upgrade-modal/index.tsx:92]: No, as I decided to refrain from using Calypso state this time. 🙂 
Related: https://github.com/Automattic/wp-calypso/pull/85013#discussion_r1421971735, https://github.com/Automattic/wp-calypso/pull/85013#discussion_r1424374607, https://github.com/Automattic/wp-calypso/pull/84886#discussion_r1426350658

File: apps/wpcom-block-editor/webpack.config.js
Comment (okmttdhr) [reply apps/wpcom-block-editor/webpack.config.js:66]: This is needed for `import`-ing `ThemeUpgradeModal`, which is directly due to the use of NODE_DEBUG in the package called [util](https://www.npmjs.com/package/util). I'll add a comment about it 👍

File: client/components/theme-upgrade-modal/index.tsx
Comment (okmttdhr) [reply client/components/theme-upgrade-modal/index.tsx:92]: CC: @renatho

**Comment by fushar** [comment]
Looks like there are conflicts... Could you resolve them and I'll test this again first thing tomorrow 😄

**Comment by okmttdhr** [comment]
Conflicts are resolved.
I'll wait for @fushar's review before merging 🙂

**Review by fushar — APPROVED** [review]
Working well.

Thanks for the hard work!

**Comment by okmttdhr** [comment]
>https://github.com/Automattic/wp-calypso/pull/85013#issuecomment-1853398541
> > > Wait a minute... So if I made some changes, then click `Upgrade now`, I will be redirected to the checkout, then got redirected back to the Site Editor, but my changes are lost... is this the expected flow? 🥲
> > > I think so because we currently don't have the mechanism to save without activating (CMIIW).
> > > But this seems weird...
> > 
> > 
> > Yes, that is expected, and we have no way to save before activation (yet) ... （´・ω・｀）.;:…（´・ω...:.;::..（´・;::: .:.;:
> 
> This seems to be a pretty significant caveat. cc: @autumnfjeld Ideally the changes are saved before leaving the Site Editor to the checkout, can we create a spike for this?

In reference to the previous discussions about saving functionality in https://github.com/Automattic/wp-calypso/issues/81119, I believe the following same principles apply here:

> The current behavior is already clear, if we think the BTP as a truly "Preview" experience (quoting @autumnfjeld's words). We also have communicated this behavior clearly in two places:
https://github.com/Automattic/wp-calypso/issues/81119#issuecomment-1831307892

Given that we can still perceive it as "a true Preview experience", and we have the notices indicating that users cannot save changes before upgrading (`You can try out your own style customizations, which will only be saved if you upgrade and activate this theme.`), 
We might want to retain user editings during the same flow (= 'Live Preview in Premium and WooCommerce themes') at least. But, if we decide to do a spike, we may want to aim to think about these complexities https://github.com/Automattic/wp-calypso/issues/81119#issuecomment-1831307892. This may include considering the saving feature in Live Preview comprehensively to ensure a consistent and non-contradictory user experience.

**Comment by taipeicoder** [comment]
> We might want to retain user editings during the same flow (= 'Live Preview in Premium and WooCommerce themes') at least.

Yes. I think this user scenario is more nuanced than https://github.com/Automattic/wp-calypso/issues/81119, in terms of (1) we are not providing a Save button to the user, and (2) I don't think it's a good UX for users to change the theme → upgrade → redo the changes → activate the theme.

The wording "You can try out your own style customizations, which will only be saved if you upgrade and activate this theme." implies the order of trying out changes → upgrade → activate, which is not technically correct since the changes will be gone after upgrading.

That said, we don't necessarily need to save when users go to checkout. Just need to improve the UX so that the second point mentioned above can be further avoided 🙂

**Comment by okmttdhr** [comment]
It would be worthwhile to see how many users make what changes before upgrading their plan in this flow before jumping in.

File: apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx
Comment (okmttdhr) [reply apps/wpcom-block-editor/src/wpcom/features/live-preview/upgrade-modal.tsx]: 📝 p7H4VZ-4JT-p2

**Pull request merged**
